File: frontend/src/api/entries.ts
import api, { API_BASE } from './client';
import type {
  JournalEntry, JournalEntryFull, EntryReadingParsed,
  FollowUpNote, Tag, Profile,
} from '../types';

// ── Entries ──────────────────────────────────────────────────

export type EntrySort = 'reading' | 'created';

export async function getEntries(
  limit = 50,
  offset = 0,
  sort: EntrySort = 'reading',
): Promise<JournalEntry[]> {
  const res = await api.get('/api/entries', { params: { limit, offset, sort } });
  return res.data;
}

export async function getEntry(entryId: number): Promise<JournalEntryFull> {
  const res = await api.get(`/api/entries/${entryId}`);
  return res.data;
}

export async function searchEntries(params: {
  query?: string;
  tag_ids?: number[];
  deck_id?: number;
  spread_id?: number;
  cartomancy_type?: string;
  card_name?: string;
  date_from?: string;
  date_to?: string;
  querent_id?: number;
  sort?: EntrySort;
}): Promise<JournalEntry[]> {
  const p: Record<string, string> = {};
  if (params.sort) p.sort = params.sort;
  if (params.query) p.query = params.query;
  if (params.tag_ids?.length) p.tag_ids = params.tag_ids.join(',');
  if (params.deck_id) p.deck_id = String(params.deck_id);
  if (params.spread_id) p.spread_id = String(params.spread_id);
  if (params.cartomancy_type) p.cartomancy_type = params.cartomancy_type;
  if (params.card_name) p.card_name = params.card_name;
  if (params.date_from) p.date_from = params.date_from;
  if (params.date_to) p.date_to = params.date_to;
  if (params.querent_id) p.querent_id = String(params.querent_id);
  const res = await api.get('/api/entries/search', { params: p });
  return res.data;
}

export async function createEntry(data: {
  title?: string;
  content?: string;
  reading_datetime?: string;
  location_name?: string;
  location_lat?: number | null;
  location_lon?: number | null;
  querent_id?: number | null;
  reader_id?: number | null;
}): Promise<{ id: number }> {
  const res = await api.post('/api/entries', data);
  return res.data;
}

export async function updateEntry(entryId: number, data: {
  title?: string;
  content?: string;
  reading_datetime?: string;
  location_name?: string;
  location_lat?: number | null;
  location_lon?: number | null;
  querent_id?: number | null;
  reader_id?: number | null;
}): Promise<void> {
  await api.put(`/api/entries/${entryId}`, data);
}

export async function deleteEntry(entryId: number): Promise<void> {
  await api.delete(`/api/entries/${entryId}`);
}

/** Persist Reading Breakdown panel state for an entry. */
export async function setEntryBreakdownSettings(
  entryId: number,
  settings: Record<string, unknown>,
): Promise<void> {
  await api.patch(`/api/entries/${entryId}/breakdown-settings`, settings);
}

// ── Readings ─────────────────────────────────────────────────

export async function getEntryReadings(entryId: number): Promise<EntryReadingParsed[]> {
  const res = await api.get(`/api/entries/${entryId}/readings`);
  return res.data;
}

export async function addEntryReading(entryId: number, data: {
  spread_id?: number | null;
  spread_name?: string;
  deck_id?: number | null;
  deck_name?: string;
  cartomancy_type?: string;
  cards_used?: Array<{ name: string; reversed?: boolean; deck_id?: number; deck_name?: string; position_index?: number }>;
  position_order?: number;
}): Promise<{ id: number }> {
  const res = await api.post(`/api/entries/${entryId}/readings`, data);
  return res.data;
}

export async function deleteEntryReadings(entryId: number): Promise<void> {
  await api.delete(`/api/entries/${entryId}/readings`);
}

/** Atomically replace all of an entry's readings in a single request.
 *  The backend performs the swap in one database transaction, so a
 *  failure can never leave the entry with its readings half-deleted.
 *  Always prefer this over deleteEntryReadings + addEntryReading loops. */
export async function replaceEntryReadings(entryId: number, readings: Array<{
  spread_id?: number | null;
  spread_name?: string;
  deck_id?: number | null;
  deck_name?: string;
  cartomancy_type?: string;
  cards_used?: Array<{ name: string; reversed?: boolean; deck_id?: number; deck_name?: string; position_index?: number; card_id?: number }>;
  position_order?: number;
}>): Promise<{ ids: number[] }> {
  const res = await api.put(`/api/entries/${entryId}/readings`, { readings });
  return res.data;
}

// ── Follow-up Notes ──────────────────────────────────────────

export async function getFollowUpNotes(entryId: number): Promise<FollowUpNote[]> {
  const res = await api.get(`/api/entries/${entryId}/follow-up-notes`);
  return res.data;
}

export async function addFollowUpNote(entryId: number, content: string): Promise<{ id: number }> {
  const res = await api.post(`/api/entries/${entryId}/follow-up-notes`, { content });
  return res.data;
}

export async function updateFollowUpNote(noteId: number, content: string): Promise<void> {
  await api.put(`/api/follow-up-notes/${noteId}`, { content });
}

export async function deleteFollowUpNote(noteId: number): Promise<void> {
  await api.delete(`/api/follow-up-notes/${noteId}`);
}

// ── Entry Tags ───────────────────────────────────────────────

export async function getEntryTags(entryId: number): Promise<Tag[]> {
  const res = await api.get(`/api/entries/${entryId}/tags`);
  return res.data;
}

export async function setEntryTags(entryId: number, tagIds: number[]): Promise<void> {
  await api.put(`/api/entries/${entryId}/tags`, { tag_ids: tagIds });
}

// ── Entry Querents ──────────────────────────────────────────

export async function setEntryQuerents(entryId: number, profileIds: number[]): Promise<void> {
  await api.put(`/api/entries/${entryId}/querents`, { profile_ids: profileIds });
}

// ── Profiles ─────────────────────────────────────────────────

export async function getProfiles(): Promise<Profile[]> {
  const res = await api.get('/api/profiles');
  return res.data;
}

// ── Export / Import ──────────────────────────────────────────

export function exportEntriesUrl(entryIds?: number[]): string {
  const params = entryIds?.length ? `?ids=${entryIds.join(',')}` : '';
  return `${API_BASE}/api/entries/export${params}`;
}

export async function importEntries(data: unknown, mergeTags = true): Promise<{
  imported: number;
  skipped: number;
  tags_created: number;
}> {
  const res = await api.post('/api/entries/import', { data, merge_tags: mergeTags });
  return res.data;
}

File: database/entries.py
"""
Database operations for journal entries, spreads, readings, and follow-up notes.
"""

import json
from datetime import datetime


class EntriesMixin:
    """Mixin providing journal entry, spread, and reading operations."""

    # === Spreads ===
    def get_spreads(self):
        cursor = self.conn.cursor()
        cursor.execute('SELECT * FROM spreads ORDER BY name')
        return cursor.fetchall()

    def get_spread(self, spread_id: int):
        cursor = self.conn.cursor()
        cursor.execute('SELECT * FROM spreads WHERE id = ?', (spread_id,))
        return cursor.fetchone()

    def add_spread(self, name: str, positions: list, description: str = None,
                   cartomancy_type: str = None, allowed_deck_types: list = None,
                   default_deck_id: int = None, deck_slots: list = None):
        """
        positions is a list of dicts: [{"x": 0, "y": 0, "label": "Past"}, ...]
        cartomancy_type: 'Tarot', 'Lenormand', 'Oracle', etc. (deprecated, for backwards compat)
        allowed_deck_types: list of cartomancy type names allowed for this spread, e.g. ['Tarot', 'Oracle']
        default_deck_id: ID of the default deck for this spread (overrides global default)
        deck_slots: list of deck slots for multi-deck spreads, e.g. [{"key": "A", "cartomancy_type": "Tarot"}]
        """
        cursor = self.conn.cursor()
        cursor.execute(
            'INSERT INTO spreads (name, description, positions, cartomancy_type, allowed_deck_types, default_deck_id, deck_slots) VALUES (?, ?, ?, ?, ?, ?, ?)',
            (name, description, json.dumps(positions), cartomancy_type,
             json.dumps(allowed_deck_types) if allowed_deck_types else None,
             default_deck_id,
             json.dumps(deck_slots) if deck_slots else None)
        )
        self._commit()
        return cursor.lastrowid

    def update_spread(self, spread_id: int, name: str = None, positions: list = None,
                      description: str = None, allowed_deck_types: list = None,
                      default_deck_id: int = None, clear_default_deck: bool = False,
                      deck_slots: list = None, archived: bool = None):
        cursor = self.conn.cursor()
        if name:
            cursor.execute('UPDATE spreads SET name = ? WHERE id = ?', (name, spread_id))
        if positions:
            cursor.execute('UPDATE spreads SET positions = ? WHERE id = ?', (json.dumps(positions), spread_id))
        if description is not None:
            cursor.execute('UPDATE spreads SET description = ? WHERE id = ?', (description, spread_id))
        if allowed_deck_types is not None:
            cursor.execute('UPDATE spreads SET allowed_deck_types = ? WHERE id = ?',
                          (json.dumps(allowed_deck_types) if allowed_deck_types else None, spread_id))
        if default_deck_id is not None or clear_default_deck:
            cursor.execute('UPDATE spreads SET default_deck_id = ? WHERE id = ?',
                          (default_deck_id, spread_id))
        if deck_slots is not None:
            cursor.execute('UPDATE spreads SET deck_slots = ? WHERE id = ?',
                          (json.dumps(deck_slots) if deck_slots else None, spread_id))
        if archived is not None:
            cursor.execute('UPDATE spreads SET archived = ? WHERE id = ?',
                          (1 if archived else 0, spread_id))
        self._commit()

    def delete_spread(self, spread_id: int):
        cursor = self.conn.cursor()
        cursor.execute('DELETE FROM spreads WHERE id = ?', (spread_id,))
        self._commit()

    # === Journal Entries ===
    # Whitelisted sort expressions for entry listings. 'reading' sorts
    # by when the reading happened (falling back to created_at for
    # entries without a reading datetime); 'created' by row creation.
    _ENTRY_SORTS = {
        'reading': 'COALESCE(reading_datetime, created_at) DESC',
        'created': 'created_at DESC',
    }

    def get_entries(self, limit: int = 50, offset: int = 0,
                    sort_by: str = 'reading'):
        order = self._ENTRY_SORTS.get(sort_by, self._ENTRY_SORTS['reading'])
        cursor = self.conn.cursor()
        # Safe dynamic SQL: `order` comes from the whitelist above.
        cursor.execute(f'''
            SELECT * FROM journal_entries
            ORDER BY {order}
            LIMIT ? OFFSET ?
        ''', (limit, offset))
        return cursor.fetchall()

    def get_entry(self, entry_id: int):
        cursor = self.conn.cursor()
        cursor.execute('SELECT * FROM journal_entries WHERE id = ?', (entry_id,))
        return cursor.fetchone()

    def search_entries(self, query: str = None, tag_ids: list = None,
                      deck_id: int = None, spread_id: int = None,
                      cartomancy_type: str = None, card_name: str = None,
                      date_from: str = None, date_to: str = None,
                      querent_id: int = None, sort_by: str = 'reading'):
        """Search entries with various filters"""
        cursor = self.conn.cursor()

        sql = 'SELECT DISTINCT je.* FROM journal_entries je'
        joins = []
        conditions = []
        params = []

        if tag_ids:
            joins.append('JOIN entry_tags et ON je.id = et.entry_id')
            # Safe IN clause: placeholders are '?' chars, values passed as params
            placeholders = ','.join('?' * len(tag_ids))
            conditions.append(f'et.tag_id IN ({placeholders})')
            params.extend(tag_ids)

        if querent_id:
            joins.append('JOIN entry_querents eq ON je.id = eq.entry_id')
            conditions.append('eq.profile_id = ?')
            params.append(querent_id)

        if deck_id or spread_id or cartomancy_type or card_name:
            joins.append('JOIN entry_readings er ON je.id = er.entry_id')
            if deck_id:
                conditions.append('er.deck_id = ?')
                params.append(deck_id)
            if spread_id:
                conditions.append('er.spread_id = ?')
                params.append(spread_id)
            if cartomancy_type:
                conditions.append('er.cartomancy_type = ?')
                params.append(cartomancy_type)
            if card_name:
                conditions.append('er.cards_used LIKE ?')
                params.append(f'%{card_name}%')

        if query:
            conditions.append('(je.title LIKE ? OR je.content LIKE ?)')
            params.extend([f'%{query}%', f'%{query}%'])

        # Date filters apply to when the READING happened (falling back
        # to created_at for entries without a reading datetime) — that's
        # the date a user remembers, not when the row was inserted.
        if date_from:
            conditions.append("COALESCE(je.reading_datetime, je.created_at) >= ?")
            params.append(date_from)

        if date_to:
            # Strictly-before-the-next-day makes a bare YYYY-MM-DD end
            # date inclusive: the old `<= date_to` compare excluded
            # everything on the end date itself, because a timestamped
            # value like '2026-07-02T10:00' sorts after '2026-07-02'.
            conditions.append("COALESCE(je.reading_datetime, je.created_at) < date(?, '+1 day')")
            params.append(date_to)

        sql += ' ' + ' '.join(joins)
        if conditions:
            sql += ' WHERE ' + ' AND '.join(conditions)
        if sort_by == 'created':
            sql += ' ORDER BY je.created_at DESC'
        else:
            sql += ' ORDER BY COALESCE(je.reading_datetime, je.created_at) DESC'

        cursor.execute(sql, params)
        return cursor.fetchall()

    def add_entry(self, title: str = None, content: str = None,
                  reading_datetime: str = None, location_name: str = None,
                  location_lat: float = None, location_lon: float = None,
                  querent_id: int = None, reader_id: int = None):
        cursor = self.conn.cursor()
        now = datetime.now().isoformat()
        # Use provided reading_datetime or default to now
        if reading_datetime is None:
            reading_datetime = now
        cursor.execute(
            '''INSERT INTO journal_entries
               (title, content, created_at, updated_at, reading_datetime, location_name, location_lat, location_lon, querent_id, reader_id)
               VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)''',
            (title, content, now, now, reading_datetime, location_name, location_lat, location_lon, querent_id, reader_id)
        )
        self._commit()
        return cursor.lastrowid

    def update_entry(self, entry_id: int, title: str = None, content: str = None,
                     reading_datetime: str = None, location_name: str = None,
                     location_lat: float = None, location_lon: float = None,
                     querent_id: int = None, reader_id: int = None):
        """Update entry fields. Safe dynamic SQL: column names are hardcoded, values use ? params."""
        cursor = self.conn.cursor()
        now = datetime.now().isoformat()
        updates = []
        params = []

        if title is not None:
            updates.append('title = ?')
            params.append(title)
        if content is not None:
            updates.append('content = ?')
            params.append(content)
        if reading_datetime is not None:
            updates.append('reading_datetime = ?')
            params.append(reading_datetime)
        if location_name is not None:
            updates.append('location_name = ?')
            params.append(location_name)
        if location_lat is not None:
            updates.append('location_lat = ?')
            params.append(location_lat)
        if location_lon is not None:
            updates.append('location_lon = ?')
            params.append(location_lon)
        if querent_id is not None:
            updates.append('querent_id = ?')
            params.append(querent_id if querent_id != 0 else None)
        if reader_id is not None:
            updates.append('reader_id = ?')
            params.append(reader_id if reader_id != 0 else None)

        if updates:
            updates.append('updated_at = ?')
            params.append(now)
            params.append(entry_id)
            cursor.execute(f'UPDATE journal_entries SET {", ".join(updates)} WHERE id = ?', params)
            self._commit()

    def set_entry_breakdown_settings(self, entry_id: int, settings_json: str):
        """Persist Reading Breakdown panel state. Doesn't bump updated_at —
        this is UI preference, not content."""
        cursor = self.conn.cursor()
        cursor.execute(
            'UPDATE journal_entries SET breakdown_settings = ? WHERE id = ?',
            (settings_json, entry_id)
        )
        self._commit()

    def delete_entry(self, entry_id: int):
        cursor = self.conn.cursor()
        cursor.execute('DELETE FROM journal_entries WHERE id = ?', (entry_id,))
        self._commit()

    # === Entry Readings ===
    def get_entry_readings(self, entry_id: int):
        cursor = self.conn.cursor()
        cursor.execute(
            'SELECT * FROM entry_readings WHERE entry_id = ? ORDER BY position_order',
            (entry_id,)
        )
        return cursor.fetchall()

    def add_entry_reading(self, entry_id: int, spread_id: int = None, spread_name: str = None,
                         deck_id: int = None, deck_name: str = None,
                         cartomancy_type: str = None, cards_used: list = None,
                         position_order: int = 0):
        cursor = self.conn.cursor()
        cursor.execute('''
            INSERT INTO entry_readings
            (entry_id, spread_id, spread_name, deck_id, deck_name, cartomancy_type, cards_used, position_order)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?)
        ''', (entry_id, spread_id, spread_name, deck_id, deck_name,
              cartomancy_type, json.dumps(cards_used) if cards_used else None, position_order))
        self._commit()
        return cursor.lastrowid

    def delete_entry_readings(self, entry_id: int):
        cursor = self.conn.cursor()
        cursor.execute('DELETE FROM entry_readings WHERE entry_id = ?', (entry_id,))
        self._commit()

    def replace_entry_readings(self, entry_id: int, readings: list) -> list:
        """Atomically replace all readings for an entry.

        The delete and re-inserts happen inside one transaction, so if
        anything fails partway (bad data, lock timeout, crash) the
        entry's original readings are left untouched instead of
        half-deleted. This is the safe path the entry editor uses when
        saving edits.

        Returns the new reading ids in order.
        """
        new_ids = []
        with self.transaction():
            cursor = self.conn.cursor()
            cursor.execute('DELETE FROM entry_readings WHERE entry_id = ?', (entry_id,))
            for i, r in enumerate(readings):
                cards_used = r.get('cards_used')
                cursor.execute('''
                    INSERT INTO entry_readings
                    (entry_id, spread_id, spread_name, deck_id, deck_name, cartomancy_type, cards_used, position_order)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
                ''', (entry_id, r.get('spread_id'), r.get('spread_name'),
                      r.get('deck_id'), r.get('deck_name'), r.get('cartomancy_type'),
                      json.dumps(cards_used) if cards_used else None,
                      r.get('position_order', i)))
                new_ids.append(cursor.lastrowid)
        return new_ids

    # === Follow-up Notes ===
    def get_follow_up_notes(self, entry_id: int):
        """Get all follow-up notes for an entry, ordered by date"""
        cursor = self.conn.cursor()
        cursor.execute('''
            SELECT * FROM follow_up_notes
            WHERE entry_id = ?
            ORDER BY created_at ASC
        ''', (entry_id,))
        return cursor.fetchall()

    def add_follow_up_note(self, entry_id: int, content: str):
        """Add a follow-up note to an entry"""
        cursor = self.conn.cursor()
        cursor.execute('''
            INSERT INTO follow_up_notes (entry_id, content, created_at)
            VALUES (?, ?, ?)
        ''', (entry_id, content, datetime.now().isoformat()))
        self._commit()
        return cursor.lastrowid

    def update_follow_up_note(self, note_id: int, content: str):
        """Update a follow-up note's content"""
        cursor = self.conn.cursor()
        cursor.execute('''
            UPDATE follow_up_notes SET content = ? WHERE id = ?
        ''', (content, note_id))
        self._commit()

    def delete_follow_up_note(self, note_id: int):
        """Delete a follow-up note"""
        cursor = self.conn.cursor()
        cursor.execute('DELETE FROM follow_up_notes WHERE id = ?', (note_id,))
        self._commit()

    # === Entry Querents ===
    def get_entry_querents(self, entry_id: int):
        """Get all querents for an entry, ordered by position"""
        cursor = self.conn.cursor()
        cursor.execute('''
            SELECT p.* FROM profiles p
            JOIN entry_querents eq ON p.id = eq.profile_id
            WHERE eq.entry_id = ?
            ORDER BY eq.position
        ''', (entry_id,))
        return cursor.fetchall()

    def set_entry_querents(self, entry_id: int, profile_ids: list):
        """Set the querents for an entry (replaces all existing)"""
        with self.transaction():
            cursor = self.conn.cursor()
            cursor.execute('DELETE FROM entry_querents WHERE entry_id = ?', (entry_id,))
            for position, profile_id in enumerate(profile_ids):
                cursor.execute('''
                    INSERT INTO entry_querents (entry_id, profile_id, position)
                    VALUES (?, ?, ?)
                ''', (entry_id, profile_id, position))
            # Also update the legacy querent_id column (first querent or NULL)
            legacy_querent_id = profile_ids[0] if profile_ids else None
            cursor.execute('''
                UPDATE journal_entries SET querent_id = ? WHERE id = ?
            ''', (legacy_querent_id, entry_id))
